Create a CLI for uploading and downloading blocks

kushalShukla-web opened the pull request:
This PR introduces a  CLI tool for seamless block uploading and downloading. The initial implementation leverages an object storage as the storage backend, with the potential for other storage backends in the future.

- Added a dedicated upload function for uploading data to the object store, with error handling and status logs using log/slog.
- A separate download function was created to fetch data from the object store.
Structured logging is a key feature of this PR, made possible by the utilization of the log/slog package. This enables easier traceability and debugging during storage operations.
- Insured clear separation of concerns by modularizing upload and download logic into distinct functions.
- Integrated the Thanos object storage package to facilitate efficient data storage and retrieval handling.

## Test plan

* [ ] Manual testing locally against a Minio service.

bboreham commented on tools/load-generator/main.go:
```diff
@@ -139,9 +139,10 @@ func (q *Querier) run(wg *sync.WaitGroup) {
 func (q *Querier) query(expr string) {
 	queryCount.WithLabelValues(q.target, q.name, expr, q.qtype).Inc()
-	start := time.Now()
+	start := time.Date(2023, time.September, 15, 10, 30, 0, 0, time.UTC)
```
This looks ok for the disk-block queries, but we will also need queries against the head-block, i.e. relative to 'now'. 
I haven't looked in detail, but suggest you make this part of the query group config.

kushalShukla-web replied:
actually i have reviste this part , i will look into this today.

bboreham reviewed (commented):
I see 'upload'; will you be downloading data at some point?

kushalShukla-web commented:
> I see 'upload'; will you be downloading data at some point?

yes @bboreham  I have to implement both upload and download functionality.

kakkoyun commented on infra/infra.go:
```diff
@@ -42,6 +43,12 @@ func main() {
 		StringMapVar(&dr.FlagDeploymentVars)
 
+		// Storing Data on Object Store
+	Objstore := app.Command("block-sync", `Using Thanos to store the data`)
```
There are a couple of things that need to be corrected here.

First of all, we were creating a new binary/CLI dedicated to block syncing. It would be better to crate a new tool in the tools directory and deploy it separately, only it's needed.

The other issue is the style here. `Objstore` should be `objstore`.

kakkoyun commented on pkg/provider/store/object_store.go:
```diff
@@ -0,0 +1,94 @@
+	ObjectConfig string
+}
+
+var S Store
```
This is not a good practice. Why do you want to have a package-level variable? Why don't we have a `NewStore` function to create a new store? We can get the bucket as a dependency and associate the function below to the store struct type.

kakkoyun reviewed (commented):
It's a good start. Let's keep up the good work.

kakkoyun commented:
@kushalShukla-web Also, we should work on the PR title and the description.

Please also sign your commits https://github.com/prometheus/test-infra/pull/760/checks?check_run_id=30633470860

kushalShukla-web committed:
Created a new Folder on tools with seperate command line

Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web commented on tools/object-storage/obj.go:
```diff
@@ -0,0 +1,110 @@
+	return err
+}
+
+func (c *Store) fun(*kingpin.ParseContext) error {
```
I will change the function name .

kakkoyun commented on tools/object-storage/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM quay.io/prometheus/busybox:latest
+LABEL maintainer="The Prometheus Authors <prometheus-developers@googlegroups.com>"
+
+COPY object-storage /bin/object-storage
```
Do you think this name describes what this tool does?

To me, it is a block syncer. It's not an object storage.

kakkoyun commented on tools/object-storage/obj.go:
```diff
@@ -0,0 +1,35 @@
+	app.HelpFlag.Short('h')
+
+	s := newstore()
+	Objstore := app.Command("block-sync", `Using Thanos to store the data`)
```
```suggestion
	objstore := app.Command("block-sync," `Using an object storage to store the data`)
```
It doesn't use the Thanos. It's using a package from Thanos. But it's an implementation detail. There's no need to leak that to the user.

kakkoyun commented on tools/object-storage/obj.go:
```diff
@@ -0,0 +1,35 @@
+
+	s := newstore()
+	Objstore := app.Command("block-sync", `Using Thanos to store the data`)
+	Objstore.Flag("path", "Path for The TSDB data in prometheus").Required().StringVar(&s.TsdbPath)
```
```suggestion
	Objstore.Flag("tsdb-path", "Path for The TSDB data in prometheus").Required().StringVar(&s.TsdbPath)
```

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+	config := c.ObjectConfig
+	configBytes, err := os.ReadFile(config)
+	if err != nil {
+		panic(fmt.Errorf("failed to read config file: %v", err))
```
This function returns an error. Why do we panic? Let's just return the error.

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+
+	bucket, err := client.NewBucket(log.NewNopLogger(), configBytes, "example")
+	if err != nil {
+		panic(err)
```
ditto.

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+	}
+	exists, err := bucket.Exists(context.Background(), "example")
+	if err != nil {
+		panic(err)
```
Same as above.

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+
+	err = objstore.UploadDir(context.Background(), log.NewNopLogger(), bucket, c.TsdbPath, commitDir)
+	if err != nil {
+		panic(err)
```
ditto.

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println(exists)
```
I would rather like to see that we use a proper logger with levels. Check out https://pkg.go.dev/log/slog

I would be happy to talk about it if you are not familiar with the structured logging concepts.

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+	}
+	exists, err := bucket.Exists(context.Background(), "example")
+	if err != nil {
+		return fmt.Errorf("failed to create bucket: %v", err)
```
You are actually returning errors here properly. Why the inconsistencies with the above error handling?

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+	ObjectConfig string
+}
+
+func newstore() *Store {
```
This constructor should consume the parameters from the command line and initialize the bucket here.

Copying from your code.
```go
    config := c.ObjectConfig
	configBytes, err := os.ReadFile(config)
	if err != nil {
		panic(fmt.Errorf("failed to read config file: %v", err))
	}
	commitDir := c.ObjectKey
	bucket, err := client.NewBucket(log.NewNopLogger(), configBytes, "example")
	if err != nil {
		panic(err)
	}
```

And you can feed the args from CLI.

kakkoyun commented on tools/object-storage/obj.go:
```diff
@@ -0,0 +1,35 @@
+	app := kingpin.New(filepath.Base(os.Args[0]), "Tool for storing TSDB data to object storage")
+	app.HelpFlag.Short('h')
+
+	s := newstore()
```
This constructor should get the flag variable. Why do we have a store in an inconsistent state?

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+	}
+	fmt.Println(exists)
+
+	err = objstore.UploadDir(context.Background(), log.NewNopLogger(), bucket, c.TsdbPath, commitDir)
```
We should give a proper logger to the buckets.

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+	}
+	commitDir := c.ObjectKey
+
+	bucket, err := client.NewBucket(log.NewNopLogger(), configBytes, "example")
```
```suggestion
	bucket, err := client.NewBucket(log.NewNopLogger(), configBytes, "block-sync")
```

We could use a proper component name.

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+	if err != nil {
+		panic(err)
+	}
+	exists, err := bucket.Exists(context.Background(), "example")
```
`example` is wrong here, no? This should be a parent directory we get as a command line argument.

kakkoyun commented on tools/object-storage/upload_download.go:
```diff
@@ -0,0 +1,88 @@
+		panic(fmt.Errorf("failed to read config file: %v", err))
+	}
+	commitDir := c.ObjectKey
+	bucket, err := client.NewBucket(log.NewNopLogger(), configBytes, "example")
```
Again we should use a proper component name.

kakkoyun reviewed (commented):
It's going in a good direction. Thanks a lot. I have commented. We need to clean up the architecture.

We should also work on the naming and documentation. We miss the README as well. 

Check this out https://go.dev/blog/package-names

kushalShukla-web committed:
Update README file and refactor code for improved readability

Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kakkoyun commented on tools/blockSyncer/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM quay.io/prometheus/busybox:latest
```
nit: It is more conventional to use `block-sync` as the directory name.

kakkoyun commented on tools/blockSyncer/Readme.md:
```diff
@@ -0,0 +1,53 @@
+
```
nit: `README.md` is more conventional.

kakkoyun commented on tools/blockSyncer/obj.go:
```diff
@@ -0,0 +1,50 @@
+	app.HelpFlag.Short('h')
+
+	var tsdbPath, objectConfig, objectKey string
+	var s *Store
```
You can omit this.

kakkoyun commented on tools/blockSyncer/obj.go:
```diff
@@ -0,0 +1,50 @@
+	objstore.Flag("tsdb-path", "Path for The TSDB data in prometheus").Required().StringVar(&tsdbPath)
+	objstore.Flag("objstore.config-file", "Path for The Config file").Required().StringVar(&objectConfig)
+	objstore.Flag("key", "Path for the Key where to store block data").Required().StringVar(&objectKey)
+
```
```suggestion

s = newstore(tsdbPath, objectConfig, objectKey)
```

kakkoyun replied:
```suggestion
s, err := newStore(...)
if err != nil {
  logger.Log(..)
 }
```

kakkoyun replied:
```suggestion
ctx, cancel := context.WithCancel(context.Backgroud())
defer cancel()
```

kakkoyun replied:
```suggestion
ctx, stop := signal.NotifyContext(ctx, os.Interrupt)
defer stop()
```

kakkoyun commented on tools/blockSyncer/obj.go:
```diff
@@ -0,0 +1,50 @@
+	objstore.Flag("tsdb-path", "Path for The TSDB data in prometheus").Required().StringVar(&tsdbPath)
+	objstore.Flag("objstore.config-file", "Path for The Config file").Required().StringVar(&objectConfig)
+	objstore.Flag("key", "Path for the Key where to store block data").Required().StringVar(&objectKey)
+
```
```suggestion
kingpin.MustParse(app.Parse(os.Args[1:]))
```

kushalShukla-web replied:
I think if we move this to line 34, it will not be able to recognize the download and upload functions because these lines are declared below.

kakkoyun commented on tools/blockSyncer/upload_download.go:
```diff
@@ -0,0 +1,90 @@
+	ObjectConfig string
+}
+
+func newstore(tsdbPath, objectConfig, objectKey string) *Store {
```
```suggestion
func newstore(tsdbPath, objectConfig, objectKey string) (*Store, error) {
```

kakkoyun commented on tools/blockSyncer/upload_download.go:
```diff
@@ -0,0 +1,90 @@
+	logger := slog.New(slog.NewJSONHandler(os.Stdout, nil))
+	configBytes, err := os.ReadFile(objectConfig)
+	if err != nil {
+		logger.Error("failed to read config file", "error", err)
```
```suggestion
		fmt.Errorf("failed to read config file, %w", err)
```

kakkoyun commented on tools/blockSyncer/upload_download.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+func newstore(tsdbPath, objectConfig, objectKey string) *Store {
+	logger := slog.New(slog.NewJSONHandler(os.Stdout, nil))
```
This needs to go to `main`.

kakkoyun commented on tools/blockSyncer/upload_download.go:
```diff
@@ -0,0 +1,90 @@
+		return nil
+	}
+
+	exists, err := bucket.Exists(context.Background(), objectKey)
```
This check needs to go to the actions.

kakkoyun commented on tools/blockSyncer/upload_download.go:
```diff
@@ -0,0 +1,90 @@
+
+func (c *Store) upload(*kingpin.ParseContext) error {
+
+	logger := slog.New(slog.NewJSONHandler(os.Stdout, nil))
```
Just initialize the logger in the constructor and store it in the `Store`.

kakkoyun commented on tools/blockSyncer/upload_download.go:
```diff
@@ -0,0 +1,90 @@
+)
+
+type Store struct {
+	Bucket       objstore.Bucket
```
```suggestion
	bucket       objstore.Bucket
```

kakkoyun commented on tools/blockSyncer/upload_download.go:
```diff
@@ -0,0 +1,90 @@
+	}
+}
+
+func (c *Store) upload(*kingpin.ParseContext) error {
```
```suggestion
func (c *Store) upload() error {
```

kakkoyun commented on tools/blockSyncer/upload_download.go:
```diff
@@ -0,0 +1,90 @@
+	return nil
+}
+
+func (c *Store) download(*kingpin.ParseContext) error {
```
```suggestion
func (c *Store) download(ctx context.Context) error {
```

kakkoyun commented on tools/blockSyncer/upload_download.go:
```diff
@@ -0,0 +1,90 @@
+
+	logger := slog.New(slog.NewJSONHandler(os.Stdout, nil))
+
+	err := objstore.UploadDir(context.Background(), log.NewNopLogger(), c.Bucket, c.TsdbPath, c.ObjectKey)
```
Create a `bucketLogger` in the constructor and use it. `c.bucketLogger`!

kushalShukla-web committed:
Good practices code

Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kakkoyun commented on tools/block-sync/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM quay.io/prometheus/busybox:latest
+LABEL maintainer="The Prometheus Authors <prometheus-developers@googlegroups.com>"
+
+COPY block-sync /bin/blocksyncer
```
```suggestion
COPY block-sync /bin/block-sync
```

Let's be consistent with the documentation.

kakkoyun commented on tools/block-sync/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+
+COPY block-sync /bin/blocksyncer
+
+ENTRYPOINT ["/bin/blocksyncer"]
```
```suggestion
ENTRYPOINT ["/bin/block-sync"]
```

kakkoyun replied:
This still needs to be addressed.

bboreham commented on tools/block-sync/README.md:
```diff
@@ -0,0 +1,53 @@
+type: s3
+config:
+  bucket: your-bucket-name
+  endpoint: https://your-minio-endpoint.com
```
More generally "your-endpoint"; it's not always minio.

kushalShukla-web committed:
Replace the library from kingpin to flag.

Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

bwplotka commented:
Is this something that would be useful in `promtool`? 🤔  https://github.com/prometheus/prometheus/tree/main/cmd/promtool

bwplotka commented:
Ah this is for old data in prombench, cool.

kushalShukla-web committed:
Implement block sync as init container for Prometheus TSDB setup

Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kakkoyun reviewed (commented):
Looking good.

kushalShukla-web committed:
Integrated block sync with Prometheus Released version and Pr version.
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

bboreham commented on prombench/manifests/prombench/benchmark/3b_prometheus-test_deployment.yaml:
```diff
@@ -144,11 +163,42 @@ spec:
+            cd /repo1 && \
+            git fetch origin pull/{{ .PR_NUMBER }}/head:pr-branch && \
+            git checkout pr-branch && \
+            cp  objectconfig.yml /mnt/repo/objectconfig.yml && \
```
This is where you would be combining password etc from some Secret?

bboreham commented on prombench/manifests/prombench/benchmark/3b_prometheus-test_deployment.yaml:
```diff
@@ -144,11 +163,42 @@ spec:
         image: quay.io/prometheus/prometheus:{{ .RELEASE }}
         imagePullPolicy: Always
-        command: [ "/bin/prometheus" ]
+        command: [ "/bin/prometheus"]
```
spurious change

bboreham commented on prombench/manifests/prombench/benchmark/3b_prometheus-test_deployment.yaml:
```diff
@@ -172,6 +222,9 @@ spec:
           path: /mnt/disks/ssd0
+      - name: config
+        hostPath:
+          path: /object-config        
```
Don't do this; just use the `instance-ssd` volume.

bboreham commented on prombench/manifests/prombench/benchmark/3b_prometheus-test_deployment.yaml:
```diff
@@ -88,6 +104,9 @@ spec:
         emptyDir: {}
+      - name: config
+        hostPath:
+          path: /config-file 
```
Don't do this; just use the `instance-ssd` volume.

bboreham commented on tools/prometheus-builder/build.sh:
```diff
@@ -6,7 +6,7 @@ if [[ -z $PR_NUMBER || -z $VOLUME_DIR || -z $GITHUB_ORG || -z $GITHUB_REPO ]]; t
     exit 1;
 fi
-
+ 
```
spurious change

bboreham commented on tools/prometheus-builder/build.sh:
```diff
@@ -24,6 +24,9 @@ fi
 git checkout pr-branch
 
+MKDIR="/config"
+cp  "$DIR/objectconfig.yml" "$MKDIR/objectconfig.yml"
```
What's going on here?

bboreham commented on tools/block-sync/README.md:
````diff
@@ -0,0 +1,53 @@
+The configuration file is essential for connecting to your object storage solution. Below are basic templates for different object storage systems.
+
+```yaml
+type: s3
````
Are other types available?  E.g. GCS?

kushalShukla-web committed:
Changed Loadgen Query to query data in respective time!
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

bboreham commented on tools/load-generator/main.go:
```diff
@@ -220,12 +227,12 @@ func main() {
 
 	var wg sync.WaitGroup
-
+	baseTime := time.Date(2024, time.October, 23, 15, 34, 32, 541000000, time.UTC)
```
eventually this should come from config, so when we upload different data we don't have to recompile the program

bboreham commented on tools/load-generator/main.go:
```diff
@@ -165,11 +166,17 @@ func (q *Querier) query(expr string) {
 	defer resp.Body.Close()
 
-	duration := time.Since(start)
+	duration := time.Since(*baseTime)
```
No

bboreham reviewed (commented):
few comments

kushalShukla-web committed:
1. Added key.yml
2. changed upload-download function to extract key
3. updated bash.sh file

Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

bboreham commented on prombench/manifests/prombench/benchmark/3b_prometheus-test_deployment.yaml:
```diff
@@ -44,10 +44,35 @@ spec:
           value: "{{ .GITHUB_ORG }}"
         - name: GITHUB_REPO
-          value: "{{ .GITHUB_REPO }}"
+          value: "{{ .GITHUB_REPO }}"     
```
spurious change

bboreham commented on prombench/manifests/prombench/benchmark/3b_prometheus-test_deployment.yaml:
```diff
@@ -34,7 +34,7 @@ spec:
       initContainers:
       - name: prometheus-builder
-        image: docker.io/prominfra/prometheus-builder:master
+        image: kushalshukla/builder
```
Guessing this one is built from the change in this PR.

bboreham commented on tools/prometheus-builder/build.sh:
```diff
@@ -5,8 +5,7 @@ DIR="/go/src/github.com/prometheus/prometheus"
     exit 1;
-fi
-
+fi 
```
Spurious change - removal of blank line.

bboreham commented on tools/prometheus-builder/build.sh:
```diff
@@ -24,6 +23,9 @@ fi
 git checkout pr-branch
 
+MKDIR="/config"
+cp  "$DIR/key.yml" "$MKDIR/key.yml"
```
This needs a comment at least.

bboreham commented on prombench/manifests/prombench/benchmark/3b_prometheus-test_deployment.yaml:
```diff
@@ -144,6 +174,46 @@ spec:
+            git fetch origin pull/{{ .PR_NUMBER }}/head:pr-branch && \
+            git checkout pr-branch && \
+            cp  key.yml /config/key.yml && \
+            rm -rf /repo1
```
This multi-line command is getting unwieldly.  Pull it out to a `.sh` file.

bboreham reviewed (commented):
We've discussed this PR on video-call and on Slack; I'm repeating a couple of points here:
* Make this PR just the CLI program; pull out the yamls and scripting to another PR.
* Get the images updated in one PR, then change the yamls that use them. We can't merge a change to use your privately-built image.

bboreham commented on prombench/manifests/prombench/benchmark/2a_prometheus-secret.yaml:
```diff
@@ -0,0 +1,16 @@
+type: Opaque
+stringData:
+  object-config.yml: |
+    type: S3
```
This is going to be configured differently for each installation - `s3` is good for local installs using Minio, but when run at Google Cloud it needs to say `gcs`.
I suggest you take this Secret out of the PR and instead put instructions in the docs, e.g. [here](https://github.com/prometheus/test-infra/blob/master/prombench/docs/gke.md), for the person who installs Prombench to set it up.

bboreham commented on prombench/manifests/prombench/benchmark/2a_prometheus-secret.yaml:
```diff
@@ -0,0 +1,16 @@
+kind: Secret
+metadata:
+  name: s3-secret
+  namespace: prombench-{{ .PR_NUMBER }}  # Replace with your actual namespace
```
I see you put the Secret in the namespace for the PR, but I see the same bucket being used across many runs.

kushalShukla-web committed:
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

Key changes
1. removed multiline command and put it in script on prometheus-builder image.
2. configure this in prometheus deployment.
3. added download.sh script which will download the data if key present , else
skip the downloading process and move to next container.

bboreham commented:
One thing that could be added (not vital right now) is to be able to upload several blocks.

Example:
```
# du -h
4.6G    ./01JC3JCA8Y39ZV72Y0F0FEEQND/chunks
5.8G    ./01JC3JCA8Y39ZV72Y0F0FEEQND
1.5G    ./01JC3J7CCY83F1EZGCBMEPG2RY/chunks
1.9G    ./01JC3J7CCY83F1EZGCBMEPG2RY
4.6G    ./01JC2XS9SGSE6YTKJZSAJP3YR8/chunks
5.8G    ./01JC2XS9SGSE6YTKJZSAJP3YR8
[...]
958.2M  ./chunks_head
85.4M   ./wal/checkpoint.00002015
4.2G    ./wal
```

I can specify one block, e.g. 01JC3J7CCY83F1EZGCBMEPG2RY, but I can't specify two.
If I upload the entire Prometheus data direectory I will get `wal` and `chunks_head`, which we don't want to download for a subsequent Prombench run.

kushalShukla-web committed:
1. Removed secretes , env variables from 3b_prometheus-test_deployment.yaml.
2. Added a section for secretes in kind.md file
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

kushalShukla-web committed:
Removed some files!

Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

bboreham reviewed (approved):
Great, let's merge this and see how well it works.
